feat: Script to migrate images between providers

FranciscoJBrito opened the pull request:
# Pull Request Template

## Description

Storage migration functionality, which allows the transfer of images from one on-premises provider to another (e.g. AWS, Google, etc.) using Active Storage.
I ran the unit and integration tests and they all passed correctly.

**The command to execute the migration is `FROM=from_service TO=to_service rake storage:migrate`**

Fixes #7907

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I have tested the storage migration functionality by running the provided unit and integration tests. Additionally, I manually tested the migration process in a local development environment by following these steps:

Set up two storage services (e.g., AWS S3 and Google Cloud Storage) with valid configurations.
Execute the `FROM=local TO=amazon rake storage:migrate` task with appropriate FROM and TO arguments to migrate blobs between the services.
Verified that the blobs were successfully transferred to the target storage service.
Checked for any error messages or unexpected behavior during the migration process.


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

FranciscoJBrito committed:
feature/#7907-script-to-change-local-storage-to-aws

FranciscoJBrito committed:
feature/#7907-script-to-change-local-storage-to-aws

FranciscoJBrito committed:
fixing path for RuboCop

FranciscoJBrito committed:
Merge branch 'chatwoot:develop' into feature/#7907-script-to-change-local-storage-to-aws

FranciscoJBrito committed:
fixing the path in the rake task test

FranciscoJBrito committed:
fixing error in the test

hiagodotme commented:
@FranciscoJBrito When trying to run on a test instance, I got the following error:

```
D, [2024-03-27T00:43:42.200427 #119] DEBUG -- : 5 Blobs to migrate from local to amazon
D, [2024-03-27T00:43:42.211579 #119] DEBUG -- : .
I, [2024-03-27T00:43:42.360708 #119]  INFO -- :   S3 Storage (148.0ms) Downloaded file from key: l7lmffqn6im34m31qupfp4wae1ww
rake aborted!
ActiveStorage::FileNotFoundError: ActiveStorage::FileNotFoundError (ActiveStorage::FileNotFoundError)
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/service/s3_service.rb:157:in `stream'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/service/s3_service.rb:41:in `block in download'
/gems/ruby/3.2.0/gems/activesupport-7.0.8.1/lib/active_support/notifications.rb:206:in `block in instrument'
/gems/ruby/3.2.0/gems/activesupport-7.0.8.1/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/gems/ruby/3.2.0/gems/activesupport-7.0.8.1/lib/active_support/notifications.rb:206:in `instrument'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/service.rb:163:in `instrument'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/service/s3_service.rb:40:in `download'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/downloader.rb:32:in `download'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/downloader.rb:13:in `block in open'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/downloader.rb:24:in `open_tempfile'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/downloader.rb:12:in `open'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/lib/active_storage/service.rb:90:in `open'
/gems/ruby/3.2.0/gems/activestorage-7.0.8.1/app/models/active_storage/blob.rb:301:in `open'
/app/lib/active_storage/migrator.rb:53:in `block in migrate_blobs'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:71:in `each'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:71:in `block in find_each'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:138:in `block in find_in_batches'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:245:in `block in in_batches'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:229:in `loop'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:229:in `in_batches'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:137:in `find_in_batches'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/relation/batches.rb:70:in `find_each'
/gems/ruby/3.2.0/gems/activerecord-7.0.8.1/lib/active_record/querying.rb:22:in `find_each'
/app/lib/active_storage/migrator.rb:48:in `migrate_blobs'
/app/lib/active_storage/migrator.rb:23:in `migrate'
/app/lib/tasks/storage_migrations.rake:9:in `block (2 levels) in <main>'
/gems/ruby/3.2.0/gems/rake-13.1.0/exe/rake:27:in `<top (required)>'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli/exec.rb:58:in `load'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli/exec.rb:58:in `kernel_load'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli/exec.rb:23:in `run'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli.rb:451:in `exec'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli.rb:34:in `dispatch'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/cli.rb:28:in `start'
/usr/local/bundle/gems/bundler-2.5.4/exe/bundle:28:in `block in <top (required)>'
/usr/local/bundle/gems/bundler-2.5.4/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
/usr/local/bundle/gems/bundler-2.5.4/exe/bundle:20:in `<top (required)>'
/usr/local/bundle/bin/bundle:25:in `load'
/usr/local/bundle/bin/bundle:25:in `<main>'
Tasks: TOP => storage:migrate
(See full trace by running task with --trace)
```

Just reporting how I did it:

First I entered my docker container:
```
$> docker exec -it rails sh
```

After that I created the files:

- /app/lib/active_storage/migrator.rb
- /app/lib/tasks/storage_migrations.rake

And I executed it as follows:

```
$> FROM=local TO=amazon bundle exec rake storage:migrate
```

I've never worked with rails, I just use CW. Did I make something wrong?

hiagodotme commented on lib/active_storage/migrator.rb:
```diff
@@ -0,0 +1,35 @@
+
+      puts "#{ActiveStorage::Blob.count} Blobs to migrate from #{from_service_name} to #{to_service_name}"
+      ActiveStorage::Blob.find_each do |blob|
+        next unless blob.image?
```
This line prevents uploading of audio files received via WhatsApp. It will possibly impact PDF's and other files in general.

FranciscoJBrito replied:
Hi @hiagodotme, of course, I assumed they wanted to export only images because the issue specifically mentioned images, but we can add other types of files without a problem. 👍🏼

FranciscoJBrito commented:
Hi @sojan-official, can I ask you for help with the checks that I am not being able to pass, if you can help me identify what the problem is why I am not being able to pass the checks???

rgon commented:
For some reason, my installation has files that cannot be found. This breaks the migration script.
<details>
<summary>Error log</summary>
<pre>
D, [2024-07-19T18:10:01.545783 #207] DEBUG -- : .
I, [2024-07-19T18:10:01.573530 #207]  INFO -- :   S3 Storage (27.0ms) Downloaded file from key: w5xh3nfsn6xt62rljwczw3avec6i
rake aborted!
ActiveStorage::FileNotFoundError: ActiveStorage::FileNotFoundError (ActiveStorage::FileNotFoundError)
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/service/s3_service.rb:157:in `stream'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/service/s3_service.rb:41:in `block in download'
/gems/ruby/3.3.0/gems/activesupport-7.0.8.4/lib/active_support/notifications.rb:206:in `block in instrument'
/gems/ruby/3.3.0/gems/activesupport-7.0.8.4/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/gems/ruby/3.3.0/gems/activesupport-7.0.8.4/lib/active_support/notifications.rb:206:in `instrument'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/service.rb:163:in `instrument'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/service/s3_service.rb:40:in `download'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/downloader.rb:32:in `download'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/downloader.rb:13:in `block in open'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/downloader.rb:24:in `open_tempfile'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/downloader.rb:12:in `open'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/lib/active_storage/service.rb:90:in `open'
/gems/ruby/3.3.0/gems/activestorage-7.0.8.4/app/models/active_storage/blob.rb:301:in `open'
/app/lib/active_storage/migrator.rb:53:in `block in migrate_blobs'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:71:in `each'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:71:in `block in find_each'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:138:in `block in find_in_batches'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:245:in `block in in_batches'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:229:in `in_batches'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:137:in `find_in_batches'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/relation/batches.rb:70:in `find_each'
/gems/ruby/3.3.0/gems/activerecord-7.0.8.4/lib/active_record/querying.rb:22:in `find_each'
/app/lib/active_storage/migrator.rb:48:in `migrate_blobs'
/app/lib/active_storage/migrator.rb:23:in `migrate'
/app/lib/tasks/storage_migrations.rake:9:in `block (2 levels) in <main>'
/gems/ruby/3.3.0/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli/exec.rb:58:in `load'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli/exec.rb:58:in `kernel_load'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli/exec.rb:23:in `run'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli.rb:455:in `exec'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli.rb:35:in `dispatch'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/cli.rb:29:in `start'
/usr/local/bundle/gems/bundler-2.5.15/exe/bundle:28:in `block in <top (required)>'
/usr/local/bundle/gems/bundler-2.5.15/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
/usr/local/bundle/gems/bundler-2.5.15/exe/bundle:20:in `<top (required)>'
/usr/local/bundle/bin/bundle:25:in `load'
/usr/local/bundle/bin/bundle:25:in `<main>'
Tasks: TOP => storage:migrate
</pre>
</details>

This issue can be solved by wrapping the `blob.open` call in `migrator.rb` in a try/catch block:

```ruby
  def self.migrate_blobs(_from_service, to_service)
    # Configure the blob service for the source service
    ActiveStorage::Blob.find_each do |blob|
      next unless blob.image?

      Rails.logger.debug { '.' }

      begin 
        blob.open do |io|
          checksum = blob.checksum
          to_service.upload(blob.key, io, checksum: checksum)
        end
      rescue ActiveStorage::FileNotFoundError => e
        # ... ignore file not found error, if it doesn't exist, it doesn't make sense to migrate it
       Rails.logger.debug { 'Non-existant file, unable to migrate it' }
      end
    end
    Rails.logger.debug { 'Successful migration' }
  end
end
```

I may have a try at committing to this PR in a while, once I manage to migrate my installation!

arthurrferroni commented:
i've made some changes in the code to works fine to run the migration.
Runned last week migrating something up to 360.000 blobs from local to AWS, all about 391gb of files.

`require 'yaml'
require 'erb'
require 'logger'

class ActiveStorage::Migrator
  Rails.logger = Logger.new($stdout)
  Rails.logger.level = Logger::DEBUG

  def self.migrate(from_service_name, to_service_name)
    configs = load_storage_config

    # Check if services are configured correctly
    if configs[from_service_name.to_s].nil? || configs[to_service_name.to_s].nil?
      raise "Erro: Os serviços '#{from_service_name}' ou '#{to_service_name}' not configured correctly."
    end

    from_service, to_service = configure_services(from_service_name, to_service_name, configs)
    configure_blob_service(from_service)

    Rails.logger.debug { "#{ActiveStorage::Blob.count} Blobs to be migrated #{from_service_name} to #{to_service_name}" }

    migrate_blobs(from_service, to_service, to_service_name)
  end

  # load the storage configs
  def self.load_storage_config
    yaml_with_env = ERB.new(File.read('config/storage.yml')).result
    YAML.load(yaml_with_env)
  end

  # Configure TO and FROM services
  def self.configure_services(from_service_name, to_service_name, configs)
    from_service = configure_service(from_service_name, configs)
    to_service = configure_service(to_service_name, configs)
    [from_service, to_service]
  end

  # Configure a single service based on settings
  def self.configure_service(service_name, configs)
    service_config = configs[service_name.to_s]
    ActiveStorage::Service.configure(service_name, { service_name.to_sym => service_config })
  end

  # Change the blob service on DB
  def self.configure_blob_service(service)
    ActiveStorage::Blob.service = service
  end

  # migrate all blobs to "TO" service
  def self.migrate_blobs(_from_service, to_service, to_service_name)
    ActiveStorage::Blob.all.each do |blob|
      migrate_blob(blob, to_service, to_service_name)
    end

    Rails.logger.debug { 'Migration Succesfully completed' }
  end

  # migrate a blob
  def self.migrate_blob(blob, to_service, to_service_name)
    # Verify if the blo is already migrated
    if blob.service_name == to_service_name.to_s
      Rails.logger.debug { "Blob #{blob.key} is already on #{to_service_name}, skipping migration." }
      return
    end

    Rails.logger.debug { "Migrating Blob #{blob.key} to#{to_service_name}" }

    begin
      blob.open do |io|
        checksum = blob.checksum
        to_service.upload(blob.key, io, checksum: checksum)
      end
      blob.update!(service_name: to_service_name.to_s)
      log_migration(blob)
    rescue ActiveStorage::FileNotFoundError
      Rails.logger.debug { "Non-existant file to blob #{blob.key}, unable to migrate it" }
    rescue StandardError => e
      handle_migration_error(blob, e)
    end
  end

  # Register the succes migrated Blob
  def self.log_migration(blob)
    Rails.logger.debug { "Blob #{blob.key} successfully migrated" }
  end

  # Register errors on migration
  def self.handle_migration_error(blob, error)
    Rails.logger.error { "Error on blob #{blob.key} migration: #{error.message}\n#{error.backtrace.join("\n")}" }
  end
end
`

arthurrferroni commented:
I made some changes to migrate blobs over providers, last week I migrated my blobs to AWS, using the code below, changes bring more readability and maintenance, I migrated more than 360000 blobs, around 391gb

```
require 'yaml'
require 'erb'
require 'logger'

class ActiveStorage::Migrator
  Rails.logger = Logger.new($stdout)
  Rails.logger.level = Logger::DEBUG

  def self.migrate(from_service_name, to_service_name)
    configs = load_storage_config

    # Check if services are configured correctly
    if configs[from_service_name.to_s].nil? || configs[to_service_name.to_s].nil?
      raise "Erro: Os serviços '#{from_service_name}' ou '#{to_service_name}' not configured correctly."
    end

    from_service, to_service = configure_services(from_service_name, to_service_name, configs)
    configure_blob_service(from_service)

    Rails.logger.debug { "#{ActiveStorage::Blob.count} Blobs to be migrated #{from_service_name} to #{to_service_name}" }

    migrate_blobs(from_service, to_service, to_service_name)
  end

  # load the storage configs
  def self.load_storage_config
    yaml_with_env = ERB.new(File.read('config/storage.yml')).result
    YAML.load(yaml_with_env)
  end

  # Configure TO and FROM services
  def self.configure_services(from_service_name, to_service_name, configs)
    from_service = configure_service(from_service_name, configs)
    to_service = configure_service(to_service_name, configs)
    [from_service, to_service]
  end

  # Configure a single service based on settings
  def self.configure_service(service_name, configs)
    service_config = configs[service_name.to_s]
    ActiveStorage::Service.configure(service_name, { service_name.to_sym => service_config })
  end

  # Change the blob service on DB
  def self.configure_blob_service(service)
    ActiveStorage::Blob.service = service
  end

  # migrate all blobs to "TO" service
  def self.migrate_blobs(_from_service, to_service, to_service_name)
    ActiveStorage::Blob.all.each do |blob|
      migrate_blob(blob, to_service, to_service_name)
    end

    Rails.logger.debug { 'Migration Succesfully completed' }
  end

  # migrate a blob
  def self.migrate_blob(blob, to_service, to_service_name)
    # Verify if the blo is already migrated
    if blob.service_name == to_service_name.to_s
      Rails.logger.debug { "Blob #{blob.key} is already on #{to_service_name}, skipping migration." }
      return
    end

    Rails.logger.debug { "Migrating Blob #{blob.key} to#{to_service_name}" }

    begin
      blob.open do |io|
        checksum = blob.checksum
        to_service.upload(blob.key, io, checksum: checksum)
      end
      blob.update!(service_name: to_service_name.to_s)
      log_migration(blob)
    rescue ActiveStorage::FileNotFoundError
      Rails.logger.debug { "Non-existant file to blob #{blob.key}, unable to migrate it" }
    rescue StandardError => e
      handle_migration_error(blob, e)
    end
  end

  # Register the succes migrated Blob
  def self.log_migration(blob)
    Rails.logger.debug { "Blob #{blob.key} successfully migrated" }
  end

  # Register errors on migration
  def self.handle_migration_error(blob, error)
    Rails.logger.error { "Error on blob #{blob.key} migration: #{error.message}\n#{error.backtrace.join("\n")}" }
  end
end

```

darker0n commented:
@arthurrferroni Hello, thank you. Your script works fine 👍

psilva-sv commented:
What's left in this PR to aprove it?

muhsin-k commented:
@sojan-official Please test these changes

sony-mathew committed:
Merge branch 'develop' into feature/#7907-script-to-change-local-storage-to-aws

sony-mathew committed:
Merge branch 'develop' into feature/#7907-script-to-change-local-storage-to-aws

sony-mathew commented:
**Verdict**

PR [[chatwoot/chatwoot#9117](https://github.com/chatwoot/chatwoot/pull/9117)](https://github.com/chatwoot/chatwoot/pull/9117) addresses a valid self-hosted need from [[#7907](https://github.com/chatwoot/chatwoot/issues/7907)](https://github.com/chatwoot/chatwoot/issues/7907): moving Active Storage blobs from local disk to S3/MinIO/etc. But it is not merge-ready. The current implementation copies some files, but does not make migrated records read from the new provider, and it also introduces a production-wide logger side effect.

**Findings**

[P1] Migrated blobs still point at the old storage service.  
[`migrator.rb`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/lib/active_storage/migrator.rb:53) opens the blob and uploads it to `to_service`, but never updates `active_storage_blobs.service_name` after success. Rails resolves blob reads through that column, and Chatwoot’s schema makes it mandatory at [`db/schema.rb:106`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/db/schema.rb:106). So after `FROM=local TO=s3_compatible rake storage:migrate`, old attachments still read from `local`; changing `ACTIVE_STORAGE_SERVICE` only affects new blobs.

[P1] The task skips non-image attachments.  
[`migrator.rb:49`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/lib/active_storage/migrator.rb:49) has `next unless blob.image?`. Chatwoot attachments include audio, video, PDFs, generic files, macro files, automation files, imports/exports, and Enterprise call/Captain documents. Users already flagged WhatsApp audio/PDF failures in the PR/issue threads. A storage migration that leaves those behind will break them once local storage is removed.

[P1] Loading the class can override production logging globally.  
[`migrator.rb:6`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/lib/active_storage/migrator.rb:6) assigns `Rails.logger = Logger.new($stdout)` and sets DEBUG level at class load time. Chatwoot eager-loads `lib` via [`config/application.rb:42`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/config/application.rb:42), and production has eager loading enabled at [`production.rb:11`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/config/environments/production.rb:11). That can override configured production logging/log level just by booting the app, before the rake task runs.

[P2] The migration is brittle for real self-hosted data.  
One missing source object will raise from `blob.open` and abort the whole run. Multiple users reported deleted/missing blobs in production installs. The task also does not filter to `service_name == FROM`, so a future version that updates `service_name` must avoid rewriting unrelated blobs.

[P2] Tests pass but don’t cover the product contract.  
[`migrator_spec.rb`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/spec/lib/active_storage/migrator_spec.rb:17) stubs `migrate_blobs`, so it never proves upload, service-name update, non-image migration, missing-file handling, or idempotency. The rake spec mutates `ENV` directly at [`task_storage_migrations_spec.rb:9`](/Users/sonymathew/.codex/worktrees/1b96/chatwoot/spec/lib/tasks/rake/task_storage_migrations_spec.rb:9) and doesn’t re-enable the task between invocations.

**Risk Audit**

No schema change, no API change, no new dependency, no feature flag, and no new external service beyond existing Active Storage providers. Security exposure is low, but operational reliability risk is high because this is a production data migration path. Enterprise has no direct overlay, but Enterprise attachments using Active Storage are impacted.

**Areas Impacted**

Self-hosted storage migrations; message attachments; contact/account avatars; portal logos; macro and automation attachments; imports/exports; Enterprise call recordings and Captain PDFs; production logging behavior.

sony-mathew committed:
fix(storage): update migrated blob service names

Default storage migrations to update active_storage_blobs.service_name after successful uploads while allowing copy-only runs with UPDATE_BLOB_SERVICE_NAME=false.

Remove unused migrator helper methods and cover the migrator/rake flag behavior in specs.

sony-mathew committed:
fix(storage): migrate all blob content types

Remove the image-only filter so storage migrations copy every Active Storage blob, including audio, documents, imports, exports, and enterprise attachments.

sony-mathew committed:
fix(storage): preserve application logger

Remove the ActiveStorage migrator class-load logger override so eager loading does not replace the configured Rails logger or log level.

sony-mathew committed:
fix(storage): make blob migration resilient

Limit storage migration to blobs on the source service, skip missing source files per blob, and keep target upload failures loud.

Clean up the service-name update flag naming and document rake task usage.

sony-mathew committed:
test(storage): cover blob migration contract

Replace mock-only migrator specs with disk-backed ActiveStorage blob coverage for non-image uploads, service-name updates, missing files, upload failures, and idempotent reruns.

Add a rake task regression for re-enabled repeated invocation.

sony-mathew commented:
The findings have been addressed with following commits:
[P1] Migrated blobs still point at the old storage service. - Fixed with [8558b6f](https://github.com/chatwoot/chatwoot/pull/9117/commits/8558b6f4c06528ec16c4d217eb44f8eced599a2a)
[P1] The task skips non-image attachments. - Fixed with [18b6681](https://github.com/chatwoot/chatwoot/pull/9117/commits/18b6681e9b52549562f04d695cf1e71852fd6b3e)
[P1] Loading the class can override production logging globally. - Fixed with [1fd280e](https://github.com/chatwoot/chatwoot/pull/9117/commits/1fd280ec0e65e7ff86a1f7d0994afe46ad02e3b6)
[P2] The migration is brittle for real self-hosted data. - fixed with [8c0cac0](https://github.com/chatwoot/chatwoot/pull/9117/commits/8c0cac0532afefa380ca067697a03de5ade58daa)
[P2] Tests pass but don’t cover the product contract. - fixed with [b2a8e03](https://github.com/chatwoot/chatwoot/pull/9117/commits/b2a8e03f62f77460d863937b567a70d2b3e617bf)

sony-mathew committed:
test(storage): isolate migrator spec service state

Restore ActiveStorage::Blob service globals after migrator specs so later attachment specs do not inherit leaked test doubles.

sony-mathew reviewed (approved):
LGTM
